ext/random/randomizer.c

```diff
@@ -258,6 +258,102 @@ PHP_METHOD(Random_Randomizer, pickArrayKeys)
 }
 /* }}} */
 
+/* {{{ Get Random Bytes for String */
+PHP_METHOD(Random_Randomizer, getBytesFromString)
+{
+	php_random_randomizer *randomizer = Z_RANDOM_RANDOMIZER_P(ZEND_THIS);
+	zend_long length;
+	zend_string *source, *retval;
+	size_t total_size = 0;
+
+	ZEND_PARSE_PARAMETERS_START(2, 2);
+		Z_PARAM_STR(source)
+		Z_PARAM_LONG(length)
+	ZEND_PARSE_PARAMETERS_END();
+
+	const size_t source_length = ZSTR_LEN(source);
+
+	if (source_length < 1) {
+		zend_argument_value_error(1, "cannot be empty");
+		RETURN_THROWS();
+	}
+
+	if (length < 1) {
+		zend_argument_value_error(2, "must be greater than 0");
+		RETURN_THROWS();
+	}
+
+	retval = zend_string_alloc(length, 0);
+
+	if (source_length > 0xFF) {
+		while (total_size < length) {
+			uint64_t offset = randomizer->algo->range(randomizer->status, 0, source_length - 1);
+
+			if (EG(exception)) {
+				zend_string_free(retval);
+				RETURN_THROWS();
+			}
+
+			ZSTR_VAL(retval)[total_size++] = ZSTR_VAL(source)[offset];
+		}
+	} else {
+		uint64_t mask;
+		if (source_length <= 0x1) {
+			mask = 0x0;
+		} else if (source_length <= 0x2) {
+			mask = 0x1;
+		} else if (source_length <= 0x4) {
+			mask = 0x3;
+		} else if (source_length <= 0x8) {
+			mask = 0x7;
+		} else if (source_length <= 0x10) {
+			mask = 0xF;
+ 		} else if (source_length <= 0x20) {
+			mask = 0x1F;
+		} else if (source_length <= 0x40) {
+			mask = 0x3F;
+		} else if (source_length <= 0x80) {
+			mask = 0x7F;
+		} else {
+			mask = 0xFF;
+		}
```

Comment:
Hmm, this isn't particularly performant. I wonder whether we should optimize that using `__builtin_clz` (and `_BitScanReverse` on Windows). We already have a few usages in php-src (see e.g. `zend_mm_small_size_to_bit()`), and it may make sense to unify these (to better encapsulate the usage of the intrinsics).

Reply:
@cmb69 I noted the same at the end of my previous comment here: https://github.com/php/php-src/pull/9664#discussion_r1030687944.

While the current if-elseif-else-construction contains several hard-to-predict branches, the performance of it is likely to be dwarfed by the loop that actually selects the random bytes.

I'd likely optimize this in a separate PR when the need arises or when there is a good API for clz. `zend_ulong_nlz` is not great, because it takes `zend_ulong` which size differs depending on the platform.